src/Features/CSharp/Portable/UsePatternCombinators/CSharpUsePatternCombinatorsAnalyzer.cs

```diff
@@ -0,0 +1,224 @@
+                case GreaterThan:
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        /// <summary>
+        /// Changes the direction the operator is pointing at.
+        /// </summary>
+        /// <remarks>
+        /// Relational patterns only come in prefix form so we'll have to
+        /// flip the operator if the original comparison has an LHS constant.
+        /// </remarks>
+        public static BinaryOperatorKind Flip(BinaryOperatorKind operatorKind)
+        {
+            return operatorKind switch
+            {
+                LessThan => GreaterThan,
+                LessThanOrEqual => GreaterThanOrEqual,
+                GreaterThanOrEqual => LessThanOrEqual,
+                GreaterThan => LessThan,
+                var v => throw ExceptionUtilities.UnexpectedValue(v)
+            };
+        }
+
+        private static bool IsConstant(IOperation operation)
+        {
+            // By-design, constants will not propagate to conversions.
+            return operation is IConversionOperation op
+                ? IsConstant(op.Operand)
+                : operation.ConstantValue.HasValue;
+        }
+
+        private static bool TryGetTargetExpression(AnalyzedPattern pattern, [NotNullWhen(true)] out ExpressionSyntax? target)
+        {
+            target = null;
+#pragma warning disable CS8762 // Parameter must have a non-null value when exiting in some condition.
+            return CheckTargetExpressions(pattern, ref target);
+#pragma warning restore CS8762 // Parameter must have a non-null value when exiting in some condition.
+
+            static bool CheckTargetExpressions(AnalyzedPattern pattern, ref ExpressionSyntax? target)
+            {
+                return pattern switch
+                {
+                    Test p => CheckTargetExpression(p.TargetOperation, ref target),
+                    Binary p => CheckTargetExpressions(p.Left, ref target) &&
+                                CheckTargetExpressions(p.Right, ref target),
+                    Not p => CheckTargetExpressions(p.Pattern, ref target),
+                    var p => throw ExceptionUtilities.UnexpectedValue(p)
+                };
+            }
+
+            static bool CheckTargetExpression(IOperation operation, ref ExpressionSyntax? target)
+            {
+                var expr = GetTargetExpression(operation);
+                if (expr is null)
+                    return false;
+
+                if (target != null)
+                    return SyntaxFactory.AreEquivalent(expr, target);
+
+                target = expr;
+                return true;
+            }
+
+            static ExpressionSyntax? GetTargetExpression(IOperation operation)
+            {
+                // Unwrap explicit casts because the pattern will emit those anyways.
+                // For instance, `(int)o == 123` would be the same as `o is 123`.
```

Comment:
i'm not sure if this is always true.  for example `(long)o == 123` is not hte same as `o is 123`.

Reply:
I've done the same thing for if-to-switch. Specially for object/string casts. I can limit this to only those types if you think it's appropriate.